### ISSUE TITLE
Re-evaluate accessibility

### ISSUE THREAD

**Issue by AdamFipke** [ubco-db/helpme#493]
For context, I have previously tried to improve the accessibility of the site for my directed studies a couple years ago. I found a lot, but we've added a bunch of features since then so it'd be good to re-evaluate everything.

Some notes:
- Pretty much don't bother with automated tools like lighthouse. They'll catch a few things, like missing `alt` attributes or bad colour contrasts, but none of the things it will point out will be *actual* accessibility issues since the things it catches are usually on elements that don't matter. These things *do* matter for stuff like SEO but we don't really care about that (so funnily enough, you can have a site with perfect 100 lighthouse accessibility scores and get bonus SEO points for it but have the site be totally inaccessible at the same time)
- Install NVDA, it's a free screen reader that seems to be one of the most popular ones, and then activate it. You can uninstall it afterwards
- As a way to start, just try navigating through the entire site with only your keyboard (arrow keys, spacebar, enter, tab, shift-tab, there's probably more hotkeys please search up and add to this list). 
  - Important things to check are stuff like forms or navigation menu elements or any other buttons. If you can't access it with the keyboard, it might as well not exist for visually impaired people which is bad. Since most of our stuff uses antd or shadcn components it *should* all be accessible by keyboard, but anything that's custom made, like a `div` with a `onClick` handler, won't be (it should be a `<button>` or `<a>` element otherwise it won't be targetable by keyboard). 
  - Check the tab order of the actions. Does it make sense? For example, when you open a modal form, does the next tab press focus the first element in the form (or close button would be fine)? Or is it focussing the `submit` button right away or is it focussing elements behind the modal?
  - Also check to make sure the screenreader is reading off anything important (and not reading off stuff it shouldn't be).
- In my original work, I added a "Skip to Main Content" button for if you're using a keyboard. Though since then I have realised that it's kinda unnecessary since there's now a built-in way of skipping to certain sections of content e.g. `<main>`  `<header>` and possibly `<h1-6>`. I never figured out what buttons to press, but for this issue I would make sure that these things work on each page. 
- I should also note that I'm unsure if all pages have a target for the "Skip to Main Content" button so that should also be checked
- One thing I noticed was that antd `message.error()` and `message.success` etc. don't notify the screen reader. I first noticed this with the update profile form, which meant that visually impaired users wouldn't know if the profile updated succeeded or failed or even went through at all. In which case I added a `sr-only` element with aria alert role, but it might be unrealistic to remember to add these elements to literally every single form and it would be better to somehow get the antd message to have the proper screen reader alert. 
  - I should also note that this was a couple years ago and antd might've already fixed this. 
  - There might be a customization that we're missing that's mentioned in the docs.
  - Maybe check the issues board for the antd repo and see if this is mentioned anywhere. If not, maybe make an issue for it on their repo
- I remember icons would get read off by the screen reader even though it's *usually* just decorative so I imagine these can be hidden with `aria-hidden` (certain buttons consist of only the icon and no text hence why this is the default). Moreover, a lot of the buttons (especially icon-only buttons) should probably have a `title` or `aria-label` attribute added to describe them better (in the case that the text of the button doesn't describe it very well).
- With the screen reader, I hadn't spent the time yet to figure out how to read paragraphs with it (might just be arrow keys?). It could be good to check this in certain parts of the application.
- I focussed almost entirely on making sure student-accessibility is good. I did not have time to evaluate TA or professor accessibility so this could be good to do (some things, like charts, probably shouldn't be bothered with).
- If you want to go deep into it, you can try navigating the site with your eyes closed and only the screen reader. Though ultimately we already know what the site looks like so we can't experience it the same way as a blind user. However, I didn't really find it necessary to close my eyes, it's usually really obvious when you can't focus specific elements or the screen reader isn't reading something it should.
